Feature: Index pages instead of tt_content

DanielSiepmann opened the pull request:
As that makes much more sense. Still content is added to each page.
Also an abstract is added, where the fields used are configurable.

DanielSiepmann committed:
FEATURE: Switch from ttcontent to pages

Also provide search_abstract as new auto added field which is
configurable.

DanielSiepmann committed:
TASK: Add documentation and config for new feature

DanielSiepmann committed:
TASK: Add tests covering new pages indexing

Including content and search abstract.

DanielSiepmann committed:
BUGFIX: Allow parent implementation to access content entry

E.g. it's possible to configure search_abstract to contain the value of
any record entry. Still content was not possible as it was set
afterwards.

DanielSiepmann committed:
TASK: Update existing tests to work with new pages setup

DanielSiepmann committed:
TASK: Check multiple fields for pages to be used as title

justusmoroni commented on Classes/Domain/Index/AbstractIndexer.php:
```diff
@@ -30,6 +32,16 @@
+    /**
+     * @var string
+     */
+    protected $identifier;
```
In other parts of code you initialized it with an empty string.
Shouldn't this be consistent?

justusmoroni commented on Classes/Domain/Index/TcaIndexer/PagesIndexer.php:
```diff
@@ -0,0 +1,101 @@
+     */
+    protected function prepareRecord(array &$record)
+    {
+        $possibleTitleFields = ['nav_title', 'tx_tqseo_pagetitle_rel', 'title'];
```
Should this be hardcoded?
We should move this to an configuration.

DanielSiepmann replied:
Yes, but for now this should be fine as two are core fields and one is just a supported extension. Once needed we can shift this to configuration with default setup.

justusmoroni commented on Configuration/TypoScript/constants.txt:
```diff
@@ -9,15 +9,9 @@ plugin {
-                        AND tt_content.CType NOT IN ('gridelements_pi1', 'list', 'div', 'menu', 'shortcut', 'search', 'login')
-                    )
+                pages {
+                    additionalWhereClause = pages.doktype NOT IN (3, 199, 6, 254, 255, 199)
```
`199` is in list multiple times.
Maybe sort it?

DanielSiepmann replied:
Sorting is done based on page type grouping, not the identifier. I've removed the second 199.

justusmoroni commented on Tests/Functional/Fixtures/Indexing/IndexTcaTable.xml:
```diff
@@ -60,4 +60,43 @@
+        <sorting>72</sorting>
+        <CType>div</CType>
+        <header>not indexed due to ctype</header>
+        <bodytext>this is the content of header content element that should not get indexed</bodytext>
```
What?

Fix in `Tests/Functional/Indexing/PagesIndexerTest.php` line 53 as well.

DanielSiepmann replied:
Nothing to fix in test, as this content element is NOT indexed due to the CType, while the header element is indexed.

But of course I've fixed the content element type in bodytext.

DanielSiepmann committed:
TASK: Cleanup code

Use get and use an try catch. Receiving null through ifExists will
result in an php error.
